### PR TITLE
feat: Add dry run flag to print support bundle specs to std out

### LINKED ISSUE #1024: Expose rendered spec after merging via an API and the cli
**Describe the rationale for the suggested feature.**

Troubleshoot has the ability to [merge multiple specs](https://troubleshoot.sh/docs/support-bundle/collecting/#collect-a-support-bundle-using-multiple-specs) and run all collectors/analysers/redactors in the resulting spec. These specs can be pulled from (as of today) the following sources
- http(s) urls
- oci uris
- files on disk (will include [multidocs as well](https://github.com/replicatedhq/troubleshoot/issues/1023))
- [embedded uris](https://troubleshoot.sh/docs/support-bundle/supportbundle/#uri)

Merged collectors are then [deduplicated](https://github.com/replicatedhq/troubleshoot/pull/972) and some default specs (cluster-resources, cluster-info, default redactors) added as well. It has now become hard to get a full picture of how the full support bundle spec would look like before executing it.

**Describe the feature**

- Define a public API that would be used to build a full spec from merged specs
- Using this public API, add a new subcommand/cli option to render the entire spec. Some examples of how the CLI interface would look like are
```sh
support-bundle render-spec <uris...> # same for preflight
support-bundle <uris...> --dry-run # same for preflight
```

**Tasks**
- [x] Add flag to `preflight` binary
- [ ] Add flag to `support-bundle` binary

### PR DESCRIPTION
**Pull request by banjoh**
## Description, Motivation and Context

Introduce `support-bundle --dry-run` flag used to print specs to stdout. The functional change is minor but I took the liberty of refactoring how we load specs from the CLI. This is where the majority of the changes lie

Fixes #1024 

**List of changes made**
- Add a new `support-bundle --dry-run` cli flag
- Have the `support-bundle` binary load specs using `LoadFromCLIArgs` API, similar to the `preflight` binary
- Mark a number of APIs as deprecated
- Make improvements to the OCI pull implementation to handle some edge cases
- Document a few APIs

**Test results**

- OCI dry run
```sh
[evans] $ support-bundle --dry-run oci://registry.replicated.com/thanos-reloaded/unstable
WARN[0002] unknown type: application/vnd.unknown.config.v1+json
WARN[0003] unknown type: application/vnd.unknown.config.v1+json
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  creationTimestamp: null
  name: merged-support-bundle-spec
spec:
  collectors:
  - clusterInfo: {}
  - clusterResources: {}
  - logs:
      namespace: '{{repl Namespace }}'
      selector:
      - app=nginx
status: {}
```
- Local path (includes support bundle and redactor)
```sh
[evans] $ support-bundle --dry-run sample-troubleshoot.yaml
apiVersion: troubleshoot.sh/v1beta2
kind: Redactor
metadata:
  creationTimestamp: null
  name: merged-redactors-spec
spec:
  redactors:
  - fileSelector:
      file: data/my-password-dump
    name: replace password
    removals:
      values:
      - abc123
  - fileSelector: {}
    name: all files
    removals:
      regex:
      - redactor: (another)(?P<mask>.*)(here)
      - redactor: '("value": ").*(")'
        selector: S3_ENDPOINT
      yamlPath:
      - abc.xyz.*
status: {}
---
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  creationTimestamp: null
  name: merged-support-bundle-spec
spec:
  collectors:
  - clusterInfo:
      collectorName: my-cluster-info
  - clusterResources:
      collectorName: my-cluster-resources
  - http:
....
```
- From url
```sh
[evans] $ support-bundle --dry-run https://raw.githubusercontent.com/replicatedhq/troubleshoot/main/sample-troubleshoot.yaml
apiVersion: troubleshoot.sh/v1beta2
kind: Redactor
metadata:
  creationTimestamp: null
  name: merged-redactors-spec
spec:
  redactors:
  - fileSelector:
      file: data/my-password-dump
    name: replace password
...
```
- From stdin
```sh
[evans] $ cat sample-troubleshoot.yaml | support-bundle --dry-run -
apiVersion: troubleshoot.sh/v1beta2
kind: Redactor
metadata:
  creationTimestamp: null
  name: merged-redactors-spec
spec:
  redactors:
  - fileSelector:
      file: data/my-password-dump
    name: replace password
....
```

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by CpuID** [comment]
I've started review this (on my phone) but need to finish on a computer - found a few things I want to verify

**Review by CpuID — COMMENTED** [review]
LGTM at a glance but there's a failing test. once the tests are passing I'll approve 👍

File: cmd/troubleshoot/cli/run.go [hunk 1]
```diff
@@ -9,35 +9,76 @@ import (
...
+		out, err := k.ToYaml()
+		if err != nil {
+			return types.NewExitCodeError(constants.EXIT_CODE_CATCH_ALL, errors.Wrap(err, "failed to convert specs to yaml"))
+		}
+		fmt.Printf("%s", out)
+		return nil
+	}
+
+	// Check if we have any collectors to run in the support bundle specs
+	if len(mainBundle.Spec.Collectors) == 0 && len(mainBundle.Spec.HostCollectors) == 0 {
```
Comment (DexterYan) [inline cmd/troubleshoot/cli/run.go]: by default, we have `ClusterInfo` and `ClusterResources` even without any collectors, I think maybe we don't need this check

Comment (banjoh) [reply cmd/troubleshoot/cli/run.go]: Good point. I moved that check to a different place.

`support-bundle` should error of there is no spec at all. Providing an empty spec should automatically add the `ClusterInfo` and `ClusterResources` collectors.

**Comment by banjoh** [comment]
> LGTM at a glance but there's a failing test. once the tests are passing I'll approve 👍

Changes causing tests to fail have been made.

**Review by CpuID — APPROVED** [review]
